sql: * in function other than COUNT causes server to panic

```
SELECT COUN(*) FROM foo
```

(for any table foo)

causes
```
panic: star qnames should really not be reaching this point!

goroutine 1218 [running]:
panic(0x15a1fc0, 0xc8204a7d80)
        /mnt/kena/go1.6.2/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/sql.multiSourceInfo.findColumn(0xc8200d4220, 0x1, 0x1, 0xc82076a280, 0x0, 0x1, 0x0, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/data_source.go:395 +0x119
github.com/cockroachdb/cockroach/sql.(*qnameVisitor).VisitPre(0xc82024fef8, 0x7fddbd355688, 0xc82076a280, 0x7fddbd355688, 0x0, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/select_qvalue.go:147 +0x18f
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7fddbd28af60, 0xc82024fef8, 0x7fddbd355688, 0xc82076a280, 0x0, 0x0, 0x1)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:415 +0x56
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Walk(0xc8203a6400, 0x7fddbd28af60, 0xc82024fef8, 0x0, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:204 +0xe6
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7fddbd28af60, 0xc82024fef8, 0x7fddbd28a5a0, 0xc8203a6400, 0x7fddbd28a5a0, 0xc8203a6400, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:418 +0x98
github.com/cockroachdb/cockroach/sql.resolveQNames(0x7fddbd28a5a0, 0xc8203a6400, 0xc8200d4220, 0x1, 0x1, 0xc82075e3c0, 0xc82024fef8, 0x0, 0x0, 0x0, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/select_qvalue.go:219 +0x17d
github.com/cockroachdb/cockroach/sql.(*planner).analyzeExpr(0xc82024f890, 0x7fddbd28a5a0, 0xc8203a6400, 0xc8200d4220, 0x1, 0x1, 0xc82075e3c0, 0x0, 0x0, 0x0, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/analyze.go:1633 +0x15d
github.com/cockroachdb/cockroach/sql.(*selectNode).addRender(0xc8201a0000, 0x7fddbd28a5a0, 0xc8203a6400, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/sql/select.go:541 +0x57c
github.com/cockroachdb/cockroach/sql.(*selectNode).initTargets(0xc8201a0000, 0xc8203b8340, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
```

--- a/sql/select_qvalue.go
+++ b/sql/select_qvalue.go
@@ -25,6 +25,7 @@ import (
 	"strings"
 
 	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/pkg/errors"
 )
 
 // columnRef is a reference to a resultColumn of a FROM node
@@ -151,14 +152,7 @@ func (v *qnameVisitor) VisitPre(expr parser.Expr) (recurse bool, newNode parser.
 		return true, v.qt.qvals.getQVal(colRef)
 
 	case *parser.FuncExpr:
-		// Special case handling for COUNT(*). This is a special construct to
-		// count the number of rows; in this case * does NOT refer to a set of
-		// columns.
-		if len(t.Name.Indirect) > 0 || !strings.EqualFold(string(t.Name.Base), "count") {
-			break
-		}
-		// The COUNT function takes a single argument. Exit out if this isn't true
-		// as this will be detected during expression evaluation.
+		// Check for invalid use of *, which, if it is an argument, is the only argument.
 		if len(t.Exprs) != 1 {
 			break
 		}
@@ -170,14 +164,18 @@ func (v *qnameVisitor) VisitPre(expr parser.Expr) (recurse bool, newNode parser.
 		if v.err != nil {
 			return false, expr
 		}
-		if !qname.IsStar() {
-			// This will cause us to recurse into the arguments of the function which
-			// will perform normal qualified name resolution.
-			break
+		if qname.IsStar() {
+			// Special case handling for COUNT(*). This is a special construct to
+			// count the number of rows; in this case * does NOT refer to a set of
+			// columns. A * is invalid elsewhere.
+			if len(t.Name.Indirect) == 0 && strings.EqualFold(string(t.Name.Base), "count") {
+				// Replace the function argument with a special non-NULL VariableExpr.
+				t = t.CopyNode()
+				t.Exprs[0] = starDatumInstance
+			} else {
+				v.err = errors.Errorf("cannot use '*' with %s", t.Name)
+			}
 		}
-		// Replace the function argument with a special non-NULL VariableExpr.
-		t = t.CopyNode()
-		t.Exprs[0] = starDatumInstance
 		return true, t
 
 	case *parser.Subquery:

--- a/sql/testdata/builtin_function
+++ b/sql/testdata/builtin_function
@@ -1064,3 +1064,7 @@ query I
 SELECT strpos(version(), 'CockroachDB')
 ----
 1
+
+# Don't panic during incorrect use of * (#7727)
+query error pq: cannot use '\*' with COS
+SELECT COS(*) FROM system.namespace

